File: lib/vision.js
/**
 * VisionIAS 100-Q Converter
 * Ported from the original HTML/JS to Node.js (Vercel serverless).
 */

const { getDocument } = require('pdfjs-dist');

async function extractPages(buffer) {
  // Node.js Buffer extends Uint8Array but pdfjs rejects it.
  // Always create a plain Uint8Array copy (no instanceof check).
  const dataArray = new Uint8Array(buffer);
  const pdf = await getDocument({ data: dataArray }).promise;
  const pages = [];
  let totalItems = 0;
  for (let p = 1; p <= pdf.numPages; p++) {
    const page = await pdf.getPage(p);
    const vp = page.getViewport({ scale: 1 });
    const tc = await page.getTextContent();
    const items = tc.items
      .filter(i => i.str && i.str.trim())
      .map(i => ({ text: i.str, x: Math.round(i.transform[4]), y: Math.round(vp.height - i.transform[5]), w: i.width, h: i.height || 10 }));
    totalItems += items.length;
    pages.push({ items, pageW: vp.width, pageH: vp.height, num: p });
  }
  if (totalItems === 0) throw new Error('No text layer found in PDF — it may be scanned/image-based.');
  return pages;
}

function itemsToLines(items, yTol = 6) {
  if (!items.length) return [];
  const sorted = [...items].sort((a, b) => a.y - b.y || a.x - b.x);
  const rows = [];
  let cur = [sorted[0]];
  for (let i = 1; i < sorted.length; i++) {
    if (Math.abs(sorted[i].y - cur[0].y) <= yTol) cur.push(sorted[i]);
    else { rows.push(cur); cur = [sorted[i]]; }
  }
  rows.push(cur);
  return rows.map(r => ({
    y: Math.round(r[0].y),
    x: Math.round(r.reduce((mn, i) => Math.min(mn, i.x), Infinity)),
    text: r.sort((a, b) => a.x - b.x).map(i => i.text).join(' ').replace(/\s{2,}/g, ' ').trim()
  })).filter(l => l.text.length > 0);
}

function pageToText(page) {
  const midX = page.pageW * 0.52;
  const leftItems = page.items.filter(i => i.x < midX);
  const rightItems = page.items.filter(i => i.x >= midX);
  const leftLines = itemsToLines(leftItems);
  const rightLines = itemsToLines(rightItems);
  if (leftLines.length >= 5 && rightLines.length >= 5) {
    return [...leftLines.map(l => l.text), ...rightLines.map(l => l.text)].join('\n');
  }
  return itemsToLines(page.items).map(l => l.text).join('\n');
}

const HF_PATTERNS = [
  /visionias/i, /vision\s*ias/i, /www\.visionias/i, /©\s*vision/i,
  /general\s+studies\s*\(p\)/i, /test\s+booklet/i,
  /answers?\s*[&and]*\s*explanations?/i,
  /^https?:\/\//i, /upscpdf\.com/i, /^www\./i, /iasscore/i,
  /time\s+allowed/i, /maximum\s+marks/i,
  /do\s+not\s+open/i, /rough\s+work/i, /invigilator/i,
  /permitted\s+to\s+take/i, /hand\s+over/i,
  /answer\s+sheet/i, /roll\s+number/i,
  /test\s+booklet\s+series/i,
  /^\d{1,3}\s*$/, /^[A-D]\s*$/,
  /IMMEDIATELY\s+AFTER/i, /ENCODE\s+CLEARLY/i,
  /this\s+test\s+booklet\s+contains\s+\d+\s+items/i,
  /you\s+have\s+to\s+(mark|enter)/i,
  /all\s+items\s+carry\s+equal/i,
  /before\s+you\s+proceed\s+to\s+mark/i,
  /after\s+you\s+have\s+completed\s+filling/i,
  /sheet\s+for\s+rough/i,
  /^\d+(st|nd|rd|th)\s*of\s*the\s*allotted/i,
  /^responses?\s*\(answers?\)/i,
  /check\s+that\s+this\s+booklet/i,
  /do\s+not\s+write\s+anything/i,
  /select\s+the\s+response/i,
  /separate\s+answer\s+sheet/i,
  /only\s+on\s+the\s*separate/i,
  /each\s+item\s+(is\s+)?printed\s+in/i,
];
function isHF(t) {
  const s = t.trim();
  return s.length === 0 || HF_PATTERNS.some(r => r.test(s));
}

const STEM_RE = /^(which\s+of\s+the|how\s+many\s+of\s+the|how\s+many\s+are|how\s+many\s+among|how\s+many\s+provisions|how\s+many\s+of\s+above|select\s+the\s+correct|choose\s+the\s+correct|in\s+how\s+many|arrange\s+the\s+following|what\s+is\s+the\s+correct|which\s+one\s+of\s+the)/i;
const SUBITEM_RE = /^(\d{1,2})\.\s+(.+)$/;
const MATCH_HEADER_RE = /^[^0-9\(].+\|.+/;
const MATCH_ROW_RE = /^(\d{1,2})\.\s+.+\|.+/;
const STMT1_RE = /^statement[- ]i\s*:/i;
const STMT2_RE = /^statement[- ]ii\s*:/i;

function isItemStem(line) {
  if (STEM_RE.test(line)) return true;
  if (/\?\s*$/.test(line)) return true;
  if (/\bhow\s+many\b/i.test(line)) return true;
  if (/select\s+the\s+correct\s+answer/i.test(line)) return true;
  return false;
}

function parseBody(rawLines) {
  let mainParts = [], subItems = [], subStem = '', matchHeader = '';
  let state = 'main', curItem = '';
  function flushItem() { const t = curItem.trim(); if (t) subItems.push(t); curItem = ''; }
  for (const raw of rawLines) {
    const line = raw.trim();
    if (!line) continue;
    const isMatchHeader = MATCH_HEADER_RE.test(line) && !SUBITEM_RE.test(line);
    const isMatchRow    = MATCH_ROW_RE.test(line);
    const sm            = line.match(SUBITEM_RE);
    const isSubItem     = !isMatchRow && sm && parseInt(sm[1]) >= 1 && parseInt(sm[1]) <= 15;
    const isStmt1       = STMT1_RE.test(line);
    const isStmt2       = STMT2_RE.test(line);
    if (state === 'main') {
      if (isMatchHeader && subItems.length === 0 && !matchHeader) { matchHeader = line; state = 'items'; }
      else if (isSubItem || isMatchRow || isStmt1) { state = 'items'; curItem = line; }
      else { mainParts.push(line); }
    } else if (state === 'items') {
      if (isItemStem(line)) { flushItem(); subStem = line; state = 'stem'; }
      else if (isMatchHeader && !matchHeader) { flushItem(); matchHeader = line; }
      else if (isSubItem || isMatchRow || isStmt1 || isStmt2) { flushItem(); curItem = line; }
      else { curItem += ' ' + line; }
    } else {
      subStem += ' ' + line;
    }
  }
  flushItem();
  let mainQ = mainParts.join(' ').replace(/\s{2,}/g, ' ').trim();
  let stem  = subStem.replace(/\s{2,}/g, ' ').trim();
  if (/how\s+many/i.test(mainQ) && subItems.length > 0 && /select\s+the\s+correct\s+answer/i.test(stem)) {
    stem = 'How many of the above are correct?';
  }
  return { mainQ, subItems, subStem: stem, matchHeader };
}

const OPT_RE  = /^\(([a-d])\)\s+(.+)$/i;
const QNUM_RE = /^(\d{1,3})\.\s+(.+)$/;

function parseTestPdf(pages, logFn) {
  const qMap = {};
  const allLines = [];
  for (const page of pages) {
    const txt = pageToText(page);
    if (!/\(\s*[abcd]\s*\)/i.test(txt)) continue;
    const ls = txt.split('\n').map(l => l.trim()).filter(l => l && !isHF(l));
    allLines.push(...ls);
  }
  if (logFn) logFn(`  Total lines after HF removal: ${allLines.length}`);

  const aIdxs = [];
  for (let i = 0; i < allLines.length; i++) {
    const m = allLines[i].match(OPT_RE);
    if (m && m[1].toLowerCase() === 'a') aIdxs.push(i);
  }
  if (logFn) logFn(`  "(a)" anchors found: ${aIdxs.length}`);

  for (let ai = 0; ai < aIdxs.length; ai++) {
    const aIdx = aIdxs[ai];
    const candidates = [];
    for (let j = aIdx - 1; j >= Math.max(0, aIdx - 80); j--) {
      const line = allLines[j];
      const om = line.match(OPT_RE);
      if (om && om[1].toLowerCase() !== 'a') break;
      const m = line.match(QNUM_RE);
      if (m) { const n = parseInt(m[1]); if (n >= 1 && n <= 100) candidates.push({ n, idx: j, text: m[2].trim() }); }
    }
    if (!candidates.length) continue;
    const { n: qNum, idx: qLineIdx, text: qFirstLine } = candidates[candidates.length - 1];
    const bodyLines = [qFirstLine];
    for (let j = qLineIdx + 1; j < aIdx; j++) {
      const line = allLines[j];
      const bm = line.match(OPT_RE);
      if (bm && bm[1].toLowerCase() !== 'a') break;
      bodyLines.push(line);
    }

    const options = [];
    let curOpt = null;
    const nextAIdx = ai + 1 < aIdxs.length ? aIdxs[ai + 1] : allLines.length;
    const scanEnd = Math.min(nextAIdx, aIdx + 40);
    for (let j = aIdx; j < scanEnd; j++) {
      const m = allLines[j].match(OPT_RE);
      if (m) {
        if (curOpt) options.push(curOpt);
        curOpt = { letter: m[1].toLowerCase(), text: m[2].trim() };
        if (m[1].toLowerCase() === 'd') { options.push(curOpt); curOpt = null; break; }
      } else if (curOpt) {
        if (QNUM_RE.test(allLines[j]) && options.length < 3) break;
        if (j !== aIdx && aIdxs.includes(j)) break;
        curOpt.text += ' ' + allLines[j];
      }
    }
    if (curOpt && !options.find(o => o.letter === curOpt.letter)) options.push(curOpt);

    if (options.length >= 2) {
      const parsed = parseBody(bodyLines);
      if (!qMap[qNum] || options.length > qMap[qNum].options.length) {
        qMap[qNum] = { parsed, options };
      }
    }
  }

  if (logFn) logFn(`  Questions parsed: ${Object.keys(qMap).length}`);
  return qMap;
}

async function parseSolPdf(pages, logFn) {
  const answers = {}, explanations = {};
  const allLines = [];
  for (const page of pages) {
    const ls = itemsToLines(page.items).filter(l => !isHF(l.text)).map(l => l.text);
    allLines.push(...ls);
  }
  const fullText = allLines.join('\n');

  const markerRe = /^Q\s*(\d{1,3})\s*\.\s*([A-D])\s*$/gm;
  const blocks = [];
  let m;
  while ((m = markerRe.exec(fullText)) !== null) {
    blocks.push({ num: parseInt(m[1], 10), letter: m[2].toLowerCase(), start: m.index, end: m.index + m[0].length });
  }

  if (blocks.length < 5) {
    if (logFn) logFn('  Trying inline answer pattern...');
    const inlineRe = /\bQ\s*(\d{1,3})\s*\.\s*([A-D])\b/g;
    while ((m = inlineRe.exec(fullText)) !== null) {
      const n = parseInt(m[1], 10);
      if (n >= 1 && n <= 100 && !blocks.find(b => b.num === n))
        blocks.push({ num: n, letter: m[2].toLowerCase(), start: m.index, end: m.index + m[0].length });
    }
    blocks.sort((a, b) => a.start - b.start);
  }

  if (logFn) logFn(`  Answer blocks found: ${blocks.length}`, blocks.length >= 90 ? 'ok' : 'warn');

  for (const b of blocks) answers[b.num] = b.letter;

  for (let i = 0; i < blocks.length; i++) {
    const { num, end } = blocks[i];
    const nextPos = i + 1 < blocks.length ? blocks[i + 1].start : fullText.length;
    const raw = fullText.slice(end, nextPos);
    const cl = cleanExplanation(raw);
    if (cl.length > 10) explanations[num] = cl;
  }

  if (logFn) logFn(`  Explanations extracted: ${Object.keys(explanations).length}`);
  return { answers, explanations };
}

function cleanExplanation(raw) {
  let t = raw;
  t = t.replace(/Hence[,\s]+option\s*\([a-d]\)\s*is\s*(the\s+)?correct\s*(answer)?\.?[^\n]*/gi, '');
  t = t.replace(/Hence\s+option\s*\(?[a-d]\)?\s*is[^.\n]*\.\s*/gi, '');
  t = t.replace(/Hence\s+option\s*\d[^.]*\.\s*/gi, '');
  t = t.replace(/Therefore[,\s]+option\s*\(?[a-d]\)?[^.]*\.\s*/gi, '');
  t = t.replace(/Hence\s+the\s+correct\s+(answer|option)[^.]*\.\s*/gi, '');
  t = t.replace(/Hence\s+statement\s+\d+\s+is\s+(correct|not\s+correct)\.\s*/gi, '');
  t = t.replace(/Hence\s+(the\s+)?statement\s+\d+\s+is[^.]*\.\s*/gi, '');
  t = t.replace(/Hence\s+option\s+\d+\s+is\s+(correct|not\s+correct)\.\s*/gi, '');
  t = t.replace(/^(Source|Note|Reference)\s*:[^\n]*/gmi, '');
  t = t.replace(/[●○•▪◆▸▹→‣]/g, '');
  t = t.replace(/^\s*o\s+/gm, ' ');
  t = t.replace(/^Q\s*\d{1,3}\s*\.\s*[A-D]\s*$/gm, '');
  const lines = t.split('\n').map(l => l.trim()).filter(l => l.length > 3 && !isHF(l));
  return lines.join(' ').replace(/\s{2,}/g, ' ').replace(/\.\s*\./g, '.').trim();
}

function buildOutput(qMap, answers, explanations) {
  const nums = Object.keys(qMap).map(Number).sort((a, b) => a - b);
  const lines = [];
  let matched = 0, noAns = 0, noExpl = 0;

  for (const num of nums) {
    const { parsed, options } = qMap[num];
    const { mainQ, subItems, subStem, matchHeader } = parsed;
    const ansLetter = answers[num];
    const expl = explanations[num] || '';

    if (!ansLetter) noAns++;
    if (!expl) noExpl++;
    if (ansLetter && expl) matched++;

    if (mainQ) {
      lines.push(`Q${num}.${mainQ}`);
      if (matchHeader) lines.push(matchHeader);
      for (const item of subItems) lines.push(item);
      if (subStem) lines.push(subStem);
    } else if (subStem) {
      lines.push(`Q${num}.${subStem}`);
      if (matchHeader) lines.push(matchHeader);
      for (const item of subItems) lines.push(item);
    } else {
      lines.push(`Q${num}.`);
    }

    lines.push('😂');

    for (const lt of ['a', 'b', 'c', 'd']) {
      const opt = options.find(o => o.letter === lt);
      if (!opt) continue;
      const txt = opt.text.replace(/\s{2,}/g, ' ').replace(/^\s*\(([a-d])\)\s*/i, '').trim();
      const mark = (ansLetter && opt.letter === ansLetter) ? ' ✅' : '';
      lines.push(txt + mark);
    }

    lines.push(expl ? `Ex: ${expl}` : `Ex: [Explanation not extracted for Q${num}]`);
    lines.push('');
  }

  return { text: lines.join('\n'), total: nums.length, matched, noAns, noExpl };
}

async function processVision(testBuffer, solBuffer, onProgress) {
  try {
    if (onProgress) onProgress(5, 'Extracting Test PDF...');
    const testPages = await extractPages(testBuffer);
    if (onProgress) onProgress(22, 'Parsing questions from Test PDF...');

    const logFn = (msg) => { if (onProgress) onProgress(22, msg); };
    const qMap = parseTestPdf(testPages, logFn);
    const qCount = Object.keys(qMap).length;

    if (qCount === 0) {
      return {
        success: false,
        error: 'No questions found in Test PDF.\nMake sure the PDF has text-based (a)(b)(c)(d) options and is not scanned.',
      };
    }

    if (qCount < 90 && onProgress) onProgress(qCount, `WARNING: Only ${qCount} questions found (expected 100)`);

    if (onProgress) onProgress(46, 'Extracting Solution PDF...');
    const solPages = await extractPages(solBuffer);

    if (onProgress) onProgress(68, 'Parsing answers and explanations...');
    const { answers, explanations } = await parseSolPdf(solPages, (msg, type) => {
      if (onProgress) onProgress(68, msg);
    });

    if (Object.keys(answers).length === 0) {
      return {
        success: false,
        error: 'No answers found in Solution PDF.\nMake sure the solution PDF has "Q 1.A" style answer headers (e.g. "Q 1.B", "Q 2.C").',
      };
    }

    if (onProgress) onProgress(88, 'Building output...');
    const { text, total, matched, noAns, noExpl } = buildOutput(qMap, answers, explanations);

    if (onProgress) onProgress(100, `Done! ${total} questions · ${matched} matched`);

    return {
      success: true,
      output: text,
      questionCount: total,
      lineCount: text.split('\n').length,
      stats: { matched, noAns, noExpl },
    };
  } catch (err) {
    return {
      success: false,
      error: err.message || String(err),
    };
  }
}

module.exports = { processVision };

File: bot/parsers/vision.py
"""
VisionIAS 100Q parser.
Requires two PDFs: Test Booklet + Answer Key / Solutions.
Answer markers in solution PDF: "Q N. Letter" (uppercase) format.
"""
import re
import pdfplumber
from io import BytesIO

# Only match clear header/footer lines — keep question/explanation content intact
NOISE_VISION = [
    re.compile(r'www\.visionias\.in', re.I),
    re.compile(r'©\s*Vision', re.I),
    re.compile(r'^\s*Test Booklet Series\s*$', re.I),
    re.compile(r'DO NOT OPEN THIS BOOKLET', re.I),
    re.compile(r'IMMEDIATELY AFTER THE COMMENCEMENT', re.I),
    re.compile(r'CANDIDATE.*RESPONSIBILITY', re.I),
    re.compile(r'ENCODE CLEARLY', re.I),
    re.compile(r'You have to enter your Roll Number', re.I),
    re.compile(r'This Test Booklet contains', re.I),
    re.compile(r'You have to mark all your responses', re.I),
    re.compile(r'All items carry equal marks', re.I),
    re.compile(r'Before you proceed to mark', re.I),
    re.compile(r'After you have completed', re.I),
    re.compile(r'Sheet for rough work', re.I),
    re.compile(r'ANSWERS\s*[&AND]+\s*EXPLANATIONS', re.I),
    re.compile(r'^\s*Time Allowed\s*:', re.I),
    re.compile(r'^\s*Maximum Marks\s*:', re.I),
    re.compile(r'^\s*\d{1,3}\s*$'),   # standalone page numbers
]

OPT_RE = re.compile(r'^\(([a-d])\)\s+(.+)$', re.I)
QNUM_RE = re.compile(r'^(\d{1,3})\.\s+(.+)$')

# Sub-item line starters in question body
NUM_STMT_RE = re.compile(r'^\d+\.\s')
STEM_RE = re.compile(
    r'^(Which|How\s+many|Select|In\s+how\s+many|Of\s+the\s+above|'
    r'From\s+the\s+above|Based\s+on)', re.I
)
MATCH_HDR_RE = re.compile(r'\s{3,}')


def is_hf(line: str) -> bool:
    s = line.strip()
    if not s:
        return True
    return any(p.search(s) for p in NOISE_VISION)


# ── PDF extraction ─────────────────────────────────────────────────────────────

def _extract_pages(pdf_bytes: bytes) -> list:
    pages = []
    with pdfplumber.open(BytesIO(pdf_bytes)) as pdf:
        for page in pdf.pages:
            words = page.extract_words(x_tolerance=3, y_tolerance=3)
            items = [{'text': w['text'], 'x': w['x0'], 'y': w['top']} for w in words]
            pages.append({
                'items': items,
                'width': float(page.width),
                'height': float(page.height),
                'page_num': page.page_number
            })
    return pages


def _items_to_lines(items: list, y_tol: int = 4) -> list:
    if not items:
        return []
    srt = sorted(items, key=lambda t: (t['y'], t['x']))
    buckets, cur = [], [srt[0]]
    for tok in srt[1:]:
        if abs(tok['y'] - cur[0]['y']) <= y_tol:
            cur.append(tok)
        else:
            buckets.append(cur)
            cur = [tok]
    buckets.append(cur)
    result = []
    for bkt in buckets:
        bkt.sort(key=lambda t: t['x'])
        txt = ' '.join(t['text'] for t in bkt).strip()
        if txt:
            result.append({'y': bkt[0]['y'], 'x': min(t['x'] for t in bkt), 'text': txt})
    return result


def _page_to_text(page: dict) -> str:
    """Extract page text handling 2-column layout (left col before right col)."""
    mid_x = page['width'] / 2
    left = [t for t in page['items'] if t['x'] < mid_x]
    right = [t for t in page['items'] if t['x'] >= mid_x]
    has_2col = len(left) > 5 and len(right) > 5

    if has_2col:
        l_lines = [l for l in _items_to_lines(left) if not is_hf(l['text'])]
        r_lines = [l for l in _items_to_lines(right) if not is_hf(l['text'])]
        return ('\n'.join(l['text'] for l in l_lines) + '\n' +
                '\n'.join(l['text'] for l in r_lines))
    else:
        all_lines = [l for l in _items_to_lines(page['items']) if not is_hf(l['text'])]
        return '\n'.join(l['text'] for l in all_lines)


# ── Body parser — matches JS parseBody logic ───────────────────────────────────

def _parse_body(body_lines: list) -> dict:
    """
    Parse question body lines into:
      mainQ      — main question text (may be empty for statement-only questions)
      subItems   — numbered sub-statements ("1. ...", "2. ...")
      subStem    — directive stem ("Which of the above...", "How many...")
      matchHeader — column header for match-type questions
    Mirrors the JavaScript parseBody() function in the original HTML.
    """
    main_parts = []
    sub_items = []
    sub_stem = ''
    match_header = None
    saw_num = False
    saw_stem = False

    for line in body_lines:
        line = line.strip()
        if not line:
            continue

        if NUM_STMT_RE.match(line):
            saw_num = True
            sub_items.append(line)
        elif STEM_RE.match(line):
            saw_stem = True
            if sub_stem:
                sub_stem += ' ' + line
            else:
                sub_stem = line
        elif MATCH_HDR_RE.search(line) and not NUM_STMT_RE.match(line) and not saw_num:
            # Spaced header line before numbered match rows
            match_header = line
        elif saw_num:
            # Continuation of last sub-item
            sub_items[-1] += ' ' + line
        elif saw_stem:
            sub_stem += ' ' + line
        else:
            main_parts.append(line)

    main_q = re.sub(r'\s{2,}', ' ', ' '.join(main_parts)).strip()
    stem = re.sub(r'\s{2,}', ' ', sub_stem).strip()

    # Normalize "How many" + "select correct" stem
    if re.search(r'how\s+many', main_q, re.I) and sub_items and re.search(r'select\s+the\s+correct', stem, re.I):
        stem = 'How many of the above are correct?'

    return {'mainQ': main_q, 'subItems': sub_items, 'subStem': stem, 'matchHeader': match_header}


# ── Test PDF parser ────────────────────────────────────────────────────────────

def _parse_test_pdf(pages: list) -> dict:
    """Anchor-on-option-a method: find (a) anchors, walk backwards to find Q number."""
    q_map = {}
    all_lines = []

    for page in pages:
        txt = _page_to_text(page)
        if not re.search(r'\(\s*[abcd]\s*\)', txt, re.I):
            continue
        ls = [l.strip() for l in txt.split('\n') if l.strip() and not is_hf(l.strip())]
        all_lines.extend(ls)

    # All indices where line is "(a) ..."
    a_idxs = [i for i, l in enumerate(all_lines)
               if OPT_RE.match(l) and OPT_RE.match(l).group(1).lower() == 'a']

    for ai, a_idx in enumerate(a_idxs):
        # Walk backwards from (a) to find question number
        candidates = []
        for j in range(a_idx - 1, max(-1, a_idx - 80), -1):
            line = all_lines[j]
            om = OPT_RE.match(line)
            if om and om.group(1).lower() != 'a':
                break   # hit a non-a option → stop
            m = QNUM_RE.match(line)
            if m:
                n = int(m.group(1))
                if 1 <= n <= 100:
                    candidates.append({'n': n, 'idx': j, 'text': m.group(2).strip()})

        if not candidates:
            continue

        # The deepest (earliest) candidate is the question number
        q_num = candidates[-1]['n']
        q_line_idx = candidates[-1]['idx']
        q_first = candidates[-1]['text']

        # Collect body lines between question number and (a)
        body_lines = [q_first]
        for j in range(q_line_idx + 1, a_idx):
            line = all_lines[j]
            bm = OPT_RE.match(line)
            if bm and bm.group(1).lower() != 'a':
                break
            body_lines.append(line)

        # Collect options (a)(b)(c)(d)
        options = []
        cur_opt = None
        next_a = a_idxs[ai + 1] if ai + 1 < len(a_idxs) else len(all_lines)
        scan_end = min(next_a, a_idx + 40)

        for j in range(a_idx, scan_end):
            m = OPT_RE.match(all_lines[j])
            if m:
                if cur_opt:
                    options.append(cur_opt)
                cur_opt = {'letter': m.group(1).lower(), 'text': m.group(2).strip()}
                if m.group(1).lower() == 'd':
                    options.append(cur_opt)
                    cur_opt = None
                    break
            elif cur_opt:
                if QNUM_RE.match(all_lines[j]) and len(options) < 3:
                    break
                if j in a_idxs and j != a_idx:
                    break
                cur_opt['text'] += ' ' + all_lines[j]

        if cur_opt and not any(o['letter'] == cur_opt['letter'] for o in options):
            options.append(cur_opt)

        if len(options) >= 2:
            parsed = _parse_body(body_lines)
            if q_num not in q_map or len(options) > len(q_map[q_num]['options']):
                q_map[q_num] = {'parsed': parsed, 'options': options}

    return q_map


# ── Solution PDF parser ────────────────────────────────────────────────────────

def _parse_sol_pdf(pages: list) -> tuple:
    """Extract answers and explanations from VisionIAS solution PDF."""
    answers = {}
    explanations = {}

    all_lines = []
    for page in pages:
        ls = [l['text'] for l in _items_to_lines(page['items']) if not is_hf(l['text'])]
        all_lines.extend(ls)

    full_text = '\n'.join(all_lines)

    # Primary: standalone "Q N. D" lines
    blocks = []
    marker_re = re.compile(r'^Q\s*(\d{1,3})\s*\.\s*([A-D])\s*$', re.M)
    for m in marker_re.finditer(full_text):
        blocks.append({'num': int(m.group(1)), 'letter': m.group(2).lower(),
                       'start': m.start(), 'end': m.end()})

    # Fallback: inline "Q1.B" or "Q 1. B" anywhere in text
    if len(blocks) < 5:
        seen = {b['num'] for b in blocks}
        inline_re = re.compile(r'\bQ\s*(\d{1,3})\s*\.\s*([A-D])\b')
        for m in inline_re.finditer(full_text):
            n = int(m.group(1))
            if 1 <= n <= 100 and n not in seen:
                blocks.append({'num': n, 'letter': m.group(2).lower(),
                               'start': m.start(), 'end': m.end()})
                seen.add(n)
        blocks.sort(key=lambda b: b['start'])

    for b in blocks:
        answers[b['num']] = b['letter']

    for i, b in enumerate(blocks):
        next_pos = blocks[i + 1]['start'] if i + 1 < len(blocks) else len(full_text)
        raw = full_text[b['end']:next_pos]
        cl = _clean_expl(raw)
        if len(cl) > 10:
            explanations[b['num']] = cl

    return answers, explanations


def _clean_expl(raw: str) -> str:
    t = raw
    t = re.sub(r'Hence[,\s]+option\s*\([a-d]\)\s*is\s*(?:the\s+)?correct\s*(?:answer)?\.?[^\n]*',
               '', t, flags=re.I)
    t = re.sub(r'Hence\s+option\s*\(?[a-d]\)?\s*is[^.\n]*\.?\s*', '', t, flags=re.I)
    t = re.sub(r'Hence\s+the\s+correct\s+(?:answer|option)[^.\n]*\.?\s*', '', t, flags=re.I)
    t = re.sub(r'Hence\s+statement\s+\d+\s+is[^.]*\.?\s*', '', t, flags=re.I)
    t = re.sub(r'Therefore[,\s]+option\s*\(?[a-d]\)?[^.]*\.?\s*', '', t, flags=re.I)
    t = re.sub(r'(?:Source|Note|Reference)\s*:[^\n]*', '', t, flags=re.I | re.M)
    t = re.sub(r'[●○•▪◆▸▹→‣]', '', t)
    t = re.sub(r'^\s*o\s+', ' ', t, flags=re.M)
    t = re.sub(r'^Q\s*\d{1,3}\s*\.\s*[A-D]\s*$', '', t, flags=re.M)
    lines = [l.strip() for l in t.split('\n') if len(l.strip()) > 3 and not is_hf(l.strip())]
    result = re.sub(r'\s{2,}', ' ', ' '.join(lines))
    return re.sub(r'\.\s*\.', '.', result).strip()


# ── Output builder ─────────────────────────────────────────────────────────────

def _build_output(q_map: dict, answers: dict, explanations: dict) -> str:
    lines_out = []
    for num in sorted(q_map.keys()):
        q = q_map[num]
        parsed = q['parsed']
        options = q['options']
        ans = answers.get(num, '')
        expl = explanations.get(num, '')

        main_q = parsed['mainQ']
        sub_items = parsed['subItems']
        sub_stem = parsed['subStem']
        match_hdr = parsed['matchHeader']

        # Question header line
        if main_q:
            lines_out.append(f'Q{num}.{main_q}')
            if match_hdr:
                lines_out.append(match_hdr)
            lines_out.extend(sub_items)
            if sub_stem:
                lines_out.append(sub_stem)
        elif sub_items or sub_stem:
            # Statement-only question (no main intro)
            stem_line = sub_stem or (sub_items[0] if sub_items else '')
            lines_out.append(f'Q{num}.{stem_line}')
            if match_hdr:
                lines_out.append(match_hdr)
            rest = sub_items[1:] if sub_stem else sub_items
            lines_out.extend(rest)
            if sub_stem and sub_items:
                pass  # stem already used as header
        else:
            lines_out.append(f'Q{num}.')

        lines_out.append('😂')

        for lt in ['a', 'b', 'c', 'd']:
            opt = next((o for o in options if o['letter'] == lt), None)
            if not opt:
                continue
            txt = re.sub(r'\s{2,}', ' ', opt['text'])
            txt = re.sub(r'^\s*\([a-d]\)\s*', '', txt, flags=re.I).strip()
            mark = ' ✅' if ans and opt['letter'] == ans else ''
            lines_out.append(txt + mark)

        lines_out.append(f'Ex: {expl}' if expl else f'Ex: [Explanation not extracted for Q{num}]')
        lines_out.append('')

    return '\n'.join(lines_out)


# ── Public API ─────────────────────────────────────────────────────────────────

def parse_vision(test_pdf_bytes: bytes, sol_pdf_bytes: bytes) -> str:
    """Parse VisionIAS test + solution PDFs. Returns formatted text."""
    test_pages = _extract_pages(test_pdf_bytes)
    q_map = _parse_test_pdf(test_pages)

    sol_pages = _extract_pages(sol_pdf_bytes)
    answers, explanations = _parse_sol_pdf(sol_pages)

    return _build_output(q_map, answers, explanations)
